feat!: no-inner-declaration new default behaviour and option (#17885)
eslint/eslint: docs/src/rules/no-inner-declarations.md

## Rule Details

> This rule requires that function declarations and, optionally, variable declarations be in the root of a program, or in the root of the body of a function, or in the root of the body of a class static block in the non-strict code. This rule doesn't report function declarations and variable declarations if code is in strict mode (code with `"use strict"` tag).

Reviewer: For clarity
```suggestion
This rule requires that function declarations and, optionally, variable declarations be in the root of a program, or in the root of the body of a function, or in the root of a class static block in non-strict code. This rule doesn't report function declarations and variable declarations if code is in strict mode (code with `"use strict"` tag or ESM modules).
```

## Options

This rule has a string and an object option:

* `"functions"` (default) disallows `function` declarations in nested blocks
* `"both"` disallows `function` and `var` declarations in nested blocks

> * `{ legacy: false }` (default) if this option set to `true` then the rule will report only function declarations even in the strict mode but only if the eslint is configured to `ecmaVersion: 5`.

Reviewer: Maybe this is clearer?
```suggestion
* `{ legacy: false }` (default) if this option set to `true` then the rule will report only function declarations when `ecmaVersion` is set to `5`.
```

> * `{ legacy: false }` (default) if this option set to `true` then the rule will report function and variable declarations (according to the string options) only if the eslint is configured to `ecmaVersion: 5`.

Reviewer 2: I don't think this option should apply to variable declarations, only function declarations.

:::

### blockScopedFunctions

Example of **incorrect** code for this rule with `{ blockScopedFunctions: "disallow" }` option with `ecmaVersion: 2015`:

::: incorrect { "sourceType": "script", "ecmaVersion": 2015 }

```js

> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "disallow" }]*/

Reviewer 2: ```suggestion
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "disallow" }]*/
```

Without the string option, this would be an invalid configuration for this rule.

// non-strict code

if (test) {
    function doSomething() { }
}

function doSomething() {
    if (test) {
        function doSomethingElse() { }
    }
}

// strict code

function foo() {
    "use strict";

    if (test) {
        function bar() { }
    }
}
```

:::

Example of **correct** code for this rule with `{ blockScopedFunctions: "disallow" }` option with `ecmaVersion: 2015`:

::: correct { "sourceType": "script", "ecmaVersion": 2015 }

```js

> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "disallow" }]*/

Reviewer 2: ```suggestion
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "disallow" }]*/
```

function doSomething() { }

function doSomething() {
    function doSomethingElse() { }
}
```

:::

Example of **correct** code for this rule with `{ blockScopedFunctions: "allow" }` option with `ecmaVersion: 2015`:

::: correct { "sourceType": "script", "ecmaVersion": 2015 }

```js

> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "allow" }]*/

Reviewer 2: ```suggestion
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "allow" }]*/
```

"use strict";

if (test) {
    function doSomething() { }
}

function doSomething() {
    if (test) {
        function doSomethingElse() { }
    }
}

// OR

function foo() {
    "use strict";

    if (test) {
        function bar() { }
    }
}
```

:::

`ESM modules` and both `class` declarations and expressions are always in strict mode.

::: correct { "sourceType": "module" }

```js

> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "allow" }]*/

Reviewer 2: ```suggestion
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "allow" }]*/
```

Author: Done!

:::

### blockScopedFunctions

Example of **incorrect** code for this rule with `{ blockScopedFunctions: "disallow" }` option with `ecmaVersion: 6`:

> ::: incorrect { "sourceType": "script" }
> 
> ```js
> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "disallow" }]*/
> /*eslint-env es6*/

Reviewer 2: ```suggestion
::: incorrect { "sourceType": "script", "ecmaVersion": 2015 }

```js
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "disallow" }]*/
```

Configuration comment was missing the string option (`["error", { blockScopedFunctions: "disallow" }]` would be an invalid configuration for this rule).

Also, `/* eslint-env */` comments do not apply anymore, but we can pass `"ecmaVersion": 2015` to the Playground.

// non-strict code

if (test) {
    function doSomething() { }
}

function doSomething() {
    if (test) {
        function doSomethingElse() { }
    }
}

// strict code

function foo() {
    "use strict";

    if (test) {
        function bar() { }
    }
}
```

:::

> Example of **correct** code for this rule with `{ blockScopedFunctions: "disallow" }` option with `ecmaVersion: 6`:

Reviewer 2: ```suggestion
Example of **correct** code for this rule with `{ blockScopedFunctions: "disallow" }` option with `ecmaVersion: 2015`:
```

> ::: correct { "sourceType": "script" }
> 
> ```js
> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "disallow" }]*/
> /*eslint-env es6*/

Reviewer 2: ```suggestion
::: correct { "sourceType": "script", "ecmaVersion": 2015 }

```js
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "disallow" }]*/
```

> "use strict";

Reviewer 2: This doesn't seem necessary in this example?

function doSomething() { }

function doSomething() {
    function doSomethingElse() { }
}
```

:::

> Example of **correct** code for this rule with `{ blockScopedFunctions: "allow" }` option with `ecmaVersion: 6`:

Reviewer 2: ```suggestion
Example of **correct** code for this rule with `{ blockScopedFunctions: "allow" }` option with `ecmaVersion: 2015`:
```

> ::: correct { "sourceType": "script" }
> 
> ```js
> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "allow" }]*/
> /*eslint-env es6*/

Reviewer 2: ```suggestion
::: correct { "sourceType": "script", "ecmaVersion": 2015 }

```js
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "allow" }]*/
```

"use strict";

if (test) {
    function doSomething() { }
}

function doSomething() {
    if (test) {
        function doSomethingElse() { }
    }
}

// OR

function foo() {
    "use strict";

    if (test) {
        function bar() { }
    }
}
```

:::

`ESM modules` and both `class` declarations and expressions are always in strict mode.

::: correct { "sourceType": "module" }

```js

> /*eslint no-inner-declarations: ["error", { blockScopedFunctions: "allow" }]*/
> /*eslint-env es6*/

Reviewer 2: ```suggestion
/*eslint no-inner-declarations: ["error", "functions", { blockScopedFunctions: "allow" }]*/
```

if (test) {
    function doSomething() { }
}

function doSomethingElse() {
    if (test) {
        function doAnotherThing() { }
    }
}

class Some {
    static {
        if (test) {
            function doSomething() { }
        }
    }
}

const C = class {
    static {
        if (test) {
            function doSomething() { }
        }
    }
}
```

:::

## When Not To Use It

> The function declaration portion rule will be rendered obsolete when [block-scoped functions](https://bugzilla.mozilla.org/show_bug.cgi?id=585536) land in ES6, but until then, it should be left on to enforce valid constructions. Disable checking variable declarations when using [block-scoped-var](block-scoped-var) or if declaring variables in nested blocks is acceptable despite hoisting.

Author: Done!

}
```

### blockScopedFunctions

:::

> Example of **correct** code for this rule with `{ blockScopedFunctions: "allow" }` option with `ecmaVersion: 6`:

Author: but please can we also keep this section to make the option more clear in `allow` ?

> Example of **correct** code for this rule with `{ blockScopedFunctions: "allow" }` option with `ecmaVersion: 6`:

Reviewer 2: Yes, makes sense to me 👍
